cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/RemoteTaskExecutorTest.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app;
+
+import com.google.common.collect.Iterators;
+import io.cdap.cdap.api.metrics.MetricValue;
+import io.cdap.cdap.api.metrics.MetricValues;
+import io.cdap.cdap.api.service.worker.RunnableTask;
+import io.cdap.cdap.api.service.worker.RunnableTaskContext;
+import io.cdap.cdap.api.service.worker.RunnableTaskRequest;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.discovery.URIScheme;
+import io.cdap.cdap.common.http.CommonNettyHttpServiceBuilder;
+import io.cdap.cdap.common.internal.remote.DefaultInternalAuthenticator;
+import io.cdap.cdap.common.internal.remote.RemoteClientFactory;
+import io.cdap.cdap.common.metrics.NoOpMetricsCollectionService;
+import io.cdap.cdap.internal.app.worker.TaskWorkerHttpHandlerInternal;
+import io.cdap.cdap.metrics.collect.AggregatedMetricsCollectionService;
+import io.cdap.cdap.security.auth.context.AuthenticationTestContext;
+import io.cdap.http.ChannelPipelineModifier;
+import io.cdap.http.NettyHttpService;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.http.HttpContentDecompressor;
+import org.apache.twill.discovery.InMemoryDiscoveryService;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Tests for RemoteTaskExecutor
+ */
+public class RemoteTaskExecutorTest {
+
+  private static RemoteClientFactory remoteClientFactory;
+  private static CConfiguration cConf;
+  private static NettyHttpService httpService;
+  private List<MetricValues> published;
+  private AggregatedMetricsCollectionService mockMetricsCollector;
+
+  public void beforeTest(boolean startHttpService) throws Exception {
```

Comment:
It's better to use `@Before` and `@After` for init and cleanup. Also, you can always have the http service started, but just cancel the discovery service registration to simulate as if the service is not available.

Reply:
Thanks. Didn't realize service registration is cancellable.